### PR TITLE
Add unimodal teacher method to train the multimodal model

### PR DESCRIPTION
**Pull request by dfulu**
# Pull Request

## Description

- Add in support for training the multimodal model via unimodal-teacher [1]
- Add tests for unimodal teacher model
- Add in support to convert the unimodal-teacher into a multimodal model and to push this to huggingface for production
- Also accidentally included another NWP / satellite encoder and tests
- Also included method to adapt batches which are too big either spatially or in time for easier experimentation
- Also light refactoring and removing duplicated code

[1] https://arxiv.org/pdf/2305.01233.pdf

## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/.github/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: pvnet/models/multimodal/encoders/basic_blocks.py [hunk 1]
```diff
@@ -78,6 +79,62 @@ def forward(self, x):
         return self.model(x) + x
 
 
+class ResidualConv3dBlock2(nn.Module):
```
Comment (jacobbieker) [inline pvnet/models/multimodal/encoders/basic_blocks.py:82]: Could we use a different name?

File: pvnet/models/multimodal/encoders/encoders3d.py [hunk 2]
```diff
@@ -201,6 +202,84 @@ def forward(self, x):
         return out
 
 
+class ResConv3DNet2(AbstractNWPSatelliteEncoder):
```
Comment (jacobbieker) [inline pvnet/models/multimodal/encoders/encoders3d.py:205]: Same here, could we use a different name? Is there some architectural thing that we could include in the name so it's not something2?

**Review by jacobbieker — APPROVED** [review]
LGTM! Great work! Only comments on naming the new architecture blocks, would be nice to differentiate them more from the original ones here.

File: pvnet/models/multimodal/encoders/encoders3d.py (hunk 2 above)
Comment (dfulu) [reply pvnet/models/multimodal/encoders/encoders3d.py:205]: The reason for this was that the structure of this is the same as the `ResFCNet2` network. And `ResFCNet2` and `ResFCNet` differed by activation functions and dropout I think. So yeh we could combine them. I'd rather do that in a second pull request than tag it on to this though
